### PR TITLE
API Add deprecations

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/silverstripe-linkfield/issues/246

### CODE REVIEW

File: src/Models/EmailLink.php [hunk 1]
```diff
@@ -18,8 +19,14 @@ class EmailLink extends Link
         'Email' => 'Varchar(255)',
     ];
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function generateLinkDescription(array $data): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/EmailLink.php]:
```suggestion
     * @deprecated 3.0.0 Use getDescription() instead.
     */
    public function generateLinkDescription(array $data): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getDescription() instead.');
```

File: src/Models/ExternalLink.php [hunk 2]
```diff
@@ -15,8 +17,14 @@ class ExternalLink extends Link
         'ExternalUrl' => 'Varchar',
     ];
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function generateLinkDescription(array $data): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/ExternalLink.php]:
```suggestion
     * @deprecated 3.0.0 Use getDescription() instead.
     */
    public function generateLinkDescription(array $data): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getDescription() instead.');
```

File: src/Models/FileLink.php [hunk 3]
```diff
@@ -18,8 +19,14 @@ class FileLink extends Link
         'File' => File::class,
     ];
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function generateLinkDescription(array $data): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/FileLink.php]:
```suggestion
     * @deprecated 3.0.0 Use getDescription() instead.
     */
    public function generateLinkDescription(array $data): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getDescription() instead.');
```
This serves the same purpose as `getDescription()`.
That method doesn't exist but this is the language we used for the 4->5 deprecations even when that method didn't exist in 4..... probably something like "Will be replaced with getDescription()" would be better though if you'd prefer that.

File: src/Models/FileLink.php [hunk 4]
```diff
@@ -31,8 +38,14 @@ public function generateLinkDescription(array $data): string
         return $file?->getFilename() ?? '';
     }
 
+    /**
+     * @deprecated 3.0.0 Will be renamed to getLinkTypeHandlerName() is linkfield v4
+     */
     public function LinkTypeHandlerName(): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be renamed to getLinkTypeHandlerName() is linkfield v4.');
```
Comment (ChloeSartorelli) [inline src/Models/FileLink.php]: Make sure you use the same language as the description method deprecation warnings, whatever wording that ends up being. Same with any other "will be renamed" type warning messages.

File: src/Models/Link.php [hunk 5]
```diff
@@ -40,30 +41,60 @@ class Link extends DataObject implements JsonData, Type
...
         return 'FormBuilderModal';
     }
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function generateLinkDescription(array $data): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/Link.php]:
```suggestion
     * @deprecated 3.0.0 Use getDescription() instead.
     */
    public function generateLinkDescription(array $data): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getDescription() instead.');
```

File: src/Models/Link.php [hunk 6]
```diff
@@ -40,30 +41,60 @@ class Link extends DataObject implements JsonData, Type
...
         return '';
     }
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function LinkTypeTile(): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/Link.php]:
```suggestion
     * @deprecated 3.0.0 Use getMenuTitle() instead.
     */
    public function LinkTypeTile(): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getMenuTitle() instead.');
```
As with the others, I don't mind what the exact wording is so long as it matches the others.

File: src/Models/Link.php [hunk 7]
```diff
@@ -243,9 +296,13 @@ public function getURL(): string
      * Get all link types except the generic one
      *
      * @throws ReflectionException
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
      */
     private function getLinkTypes(): array
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/Link.php]:
```suggestion
     * @deprecated 3.0.0 Use SilverStripe\LinkField\Services\LinkTypeService::generateAllLinkTypes() instead.
     */
    private function getLinkTypes(): array
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use SilverStripe\LinkField\Services\LinkTypeService::generateAllLinkTypes() instead.');
```
Again, exact wording so long as it matches the others - main point is this has a replacement in v4

File: src/Models/PhoneLink.php [hunk 8]
```diff
@@ -15,8 +17,14 @@ class PhoneLink extends Link
         'Phone' => 'Varchar(255)',
     ];
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function generateLinkDescription(array $data): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/PhoneLink.php]:
```suggestion
     * @deprecated 3.0.0 Use getDescription() instead.
     */
    public function generateLinkDescription(array $data): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getDescription() instead.');
```

File: src/Models/SiteTreeLink.php [hunk 9]
```diff
@@ -30,8 +31,14 @@ class SiteTreeLink extends Link
         'Page' => SiteTree::class,
     ];
 
+    /**
+     * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
+     */
     public function generateLinkDescription(array $data): string
     {
+        Deprecation::withNoReplacement(function () {
+            Deprecation::notice('3.0.0', 'Will be removed without equivalent functionality to replace it.');
```
Comment (ChloeSartorelli) [inline src/Models/SiteTreeLink.php]:
```suggestion
     * @deprecated 3.0.0 Use getDescription() instead.
     */
    public function generateLinkDescription(array $data): string
    {
        Deprecation::withNoReplacement(function () {
            Deprecation::notice('3.0.0', 'Use getDescription() instead.');
```

File: src/Type/Registry.php [hunk 10]
```diff
@@ -8,16 +8,26 @@
 use SilverStripe\Core\Extensible;
 use SilverStripe\Core\Injector\Injectable;
 use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Dev\Deprecation;
 
 /**
  * Manage the list of known link types
+ *
+ * @deprecated 3.0.0 Will be removed without equivalent functionality to replace it
```
Comment (ChloeSartorelli) [inline src/Type/Registry.php]: I think this was renamed to `LinkTypeService` - some of the same methods are there (e.g. `byKey()`)

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Mostly just a few methods that do have replacements. I've suggested some wording but feel free to word it differently than my suggestions so long as it's consistent.

File: src/Type/Registry.php (hunk 10 above)
Comment (emteknetnz) [reply src/Type/Registry.php]: Updated

File: src/Models/EmailLink.php (hunk 1 above)
Comment (emteknetnz) [reply src/Models/EmailLink.php]: Updated

File: src/Models/ExternalLink.php (hunk 2 above)
Comment (emteknetnz) [reply src/Models/ExternalLink.php]: Updated

File: src/Models/FileLink.php (hunk 3 above)
Comment (emteknetnz) [reply src/Models/FileLink.php]: Updated

File: src/Models/FileLink.php (hunk 4 above)
Comment (emteknetnz) [reply src/Models/FileLink.php]: Updated

File: src/Models/Link.php (hunk 5 above)
Comment (emteknetnz) [reply src/Models/Link.php]: Updated

File: src/Models/Link.php (hunk 6 above)
Comment (emteknetnz) [reply src/Models/Link.php]: Updated

File: src/Models/Link.php (hunk 7 above)
Comment (emteknetnz) [reply src/Models/Link.php]: Updated

File: src/Models/PhoneLink.php (hunk 8 above)
Comment (emteknetnz) [reply src/Models/PhoneLink.php]: Updated

File: src/Models/SiteTreeLink.php (hunk 9 above)
Comment (emteknetnz) [reply src/Models/SiteTreeLink.php]: Updated

**Review by GuySartorelli — APPROVED** [review]
LGTM

**Pull request merged**